redis.conf

```diff
@@ -1574,6 +1574,13 @@ lua-time-limit 5000
 #
 # cluster-allow-reads-when-down no
 
+# Cluster link send buffer limit is the limit on the memory usage of an individual
+# cluster bus link's send buffer in bytes. Cluster links would be freed if they exceed
+# this limit. This is to primarily prevent send buffers from growing unbounded on links
+# toward slow peers (E.g. PubSub messages being piled up).
+#
+# cluster-link-sendbuf-limit 1gb
```

Comment:
I forgot where we came up with 1GB, but lowering seems reasonable. I think the limit is somewhere in between the two. We expect the other side to be consuming the data as fast as possible, like a replica, but it also has the high write amplification like pubsub. Since each individual link is small, I agree that 32mb sounds reasonable.

Reply:
This feature aims to protect against any case where a peer cannot drain cluster link send buffers as fast as Redis can fill them, e.g. hang peer, slow peer, etc.

I chose 1GB as the default limit based on the max possible size of a single PubSub command, which is 1GB by default (`client-query-buffer-limit`'s default value). I wanted cluster link buffer to be able to fit at least a single PubSub command by default.

Reply:
i suppose that effectively this means for many deployments as if the default is practically unlimited (will still cause mass eviction / OOM).
maybe that's a good thing, to consider that as if it's not a "breaking change", i.e. the new default has the same behavior as the old versions, but if we look at it like that, maybe just set it to 0 by default?

if we have any specific reasoning like the one you just mentioned, maybe we should document it in the redis.conf and also in the top comment.

Reply:
I suppose one problem with 32MB is that pubsub messages might get constantly dropped and it would be very hard to notice. Unlike a regular pubsub client, which will get repeatedly disconnected if large messages are put in it, the cluster link will just get dropped and re-established, but the "connected" pubsub client won't notice anything. So I guess I'm back in the 1gb camp, and we'll use it more as a mechanism to make sure it doesn't grow unbounded.

Reply:
I was arguing that for **some** deployments it's practically unlimited. I.e. Ones with not a lot of memory (either physical, or just the maxmemory config), or ones with a lot of cluster nodes. 

So considering that, i was arguing that maybe the default should be disabled, and thus not changing default behaviour on upgrades. 

P.s. Note that nearly all the memory limit setting in Redis are off by default (except for client buffers). This is because it's impossible to come up with a number that will be a good default for everyone. 

Anyway, considering what @madolson wrote, about getting disconnections and reconnection, resulting the silent loss of messages with no clear issue indicated anywhere, I'm not sure i like that feature at all..

Reply:
I hear you about it's impossible to pick a default to make everyone happy.

I also agree that **silently** dropping cluster links and PubSub broadcast messages is bad.

Please allow me to re-visit why I wrote this PR in the first place. Before this PR, cluster link buffers can grow unbounded and:
1. It's impossible to observe if it is happening
2. It's impossible to isolate which link is using memory
3. It's impossible to mitigate it with minimal impact on the server (there is no easy way to free a specific cluster link that is over-using memory).

In this PR:
1. `mem_cluster_links` INFO field is introduced to address (1)
2. `CLUSTER LINKS` command is introduced to address (2)
3. `cluster-link-sendbuf-limit` config is introduced to address (3).

In this context, re: `enabling cluster-link-sendbuf-limit by default or not`. I was leaning toward enabling by default. Because I trust the default of 1GB will not produce "false alarms". That is, we never expect a cluster link to use >1GB by default. When a cluster link does, it means Redis is operating in an abnormal and degraded mode, which we want to get out of. Enabling by default just means Redis would auto-mitigate such an abnormal situation, which is a net positive. On smaller deployments where there is not a lot of memory, this auto-mitigation will never take effect unfortunately, so Redis regresses to its current behavior, which is not a negative thing either.

Re: `Using 1GB as default`. I guess my reasonings can be summarized as:
1. In conjunction with `client-query-buffer-limit'`s default value of 1GB, a default 1GB of `cluster-link-sendbuf-limit` should/would not produce false alarms, which is important.
2. While it's impossible to pick a default everyone likes, sometimes we still have to pick a sensible number to give us some bottom line protection. E.g. `cluster-link-sendbuf-limit` default to 1GB is a good example. I tried to follow that.

Re: `silently dropping cluster links and PubSub messages`. I agree this is bad. This is a separate issue. It seems like whenever `cluster-link-sendbuf-limit` takes effect and frees a link, we also want to print a log and/or increment a metric count (e.g. a `cluster_links_freed_oom` INFO field, etc). So it's not "silent". I can add that in the next revision if we all agree to if we want a log, a metric, or both.

Thank you for your feedback. Please let me know your thoughts.

Reply:
I'm still in favor of setting 1gb, but only making this available in Redis 7. In some sense it is a backward breaking change, but it is bounding the problem moving forward.

Reply:
Decision by core team was to set the default to 0.

Reply:
Done.